### PR TITLE
Add fullbright mode

### PR DESCRIPTION
**Pull request by HybridDog**
It works for nodes (solid and not solid (e.g. plantlike)), objects and particles with enabled or disabled smooth lighting, did I omit something?
I don't know how to efficiently pass the fullbright setting (which is changed now and then) to the light calculation (which happens constantly) and I don't know whether the mapblock mesh needs to be recalculated when toggling the setting, so I cannot finish this PR.
I've made two commits, one for the full brightness and another one for the key, privilege and similar things.
Note: Adding fullbright mode is in Things to do soon on the [official TODO list](http://dev.minetest.net/TODO#Other).
See also #6509

### CODE REVIEW

**Comment by sfan5** [comment]
I'm not sure if this is really needed, can't this be put under the `debug` privilege.

**Comment by Ezhh** [comment]
I have wondered whether something like this would be good to have, but always end up unsure. On my creative server I have many builders who prefer to build in the light, but  it's also important to be able to check how something appears in the dark, so this would only be useful in such situations if it's a separate priv and if players are still able to experience darkness when they want to.

Alternatively it could be useful for testing mods, but despite working on several mods for underground areas, this feels like a less important use to me.

**Comment by HybridDog** [comment]
What does the debug privilege allow and prohibit?
#
> […] builders who prefer to build in the light […]

I think lots of players want to build in the light and place torches in the end. It's quite inconvenient when you constantly have to place and dig torches just to see what you are building.

> […] if players are still able to experience darkness when they want to.

Press L to toggle fullbright mode.

**Comment by Ezhh** [comment]
As long it's it can be toggled, I would be more than happy to see this added.

Debug priv gives the wireframe mode, and maybe other info as well. I'd be happy to allow many players on my server fullbright, but I wouldn't want to give them wireframe mode, so for me that would be a bad priv to tack this on to. Fullbright isn't really about debugging (or certainly not only about debugging).

**Comment by stujones11** [comment]
I can see this as a potential game feature, night-vision glasses, for example. Maybe this would be better as a per-player property like `zoom_fov`?

**Comment by NewbProgrammer101** [comment]
👍

**Comment by Fixer-007** [comment]
night-vision?

**Comment by stujones11** [comment]
> night-vision?

Infrared, could also be used for guns, etc. You would probably want to combine it with an opaque HUD filter to maximise the effect :)

**Comment by mixrat** [comment]
How about an API to override client settings in order to be used for night vision goggles so player doesn't need to use 'L' when wearing them?

@Wuzzy you just need to compile a test client with `light_source = 15;` @ line 378 of src/nodedef.cpp.

**Comment by ClobberXD** [comment]
> How about an API to override client settings in order to be used for night vision goggles so player doesn't need to use 'L' when wearing them?

I agree, (only) mods should be able to toggle full-bright mode for players.

**Comment by paramat** [comment]
> How about an API to override client settings in order to be used for night vision goggles

Fullbright is completely unsuitable for 'night vision' etc., it's only suitable for development / creative building / admin uses.
So this seems best as some kind of client visual mode, it doesn't need an API for mod control.

File: builtin/settingtypes.txt [hunk 1]
```diff
@@ -241,6 +241,11 @@ keymap_cinematic (Cinematic mode key) key
 #    See http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3
 keymap_minimap (Minimap key) key KEY_F9
 
+#    Key for toggling fullbright mode.
+#    It's set to the key l by default because l is the first letter of light.
```
Comment (paramat) [inline builtin/settingtypes.txt]: This comment is unnecessary.

File: builtin/settingtypes.txt [hunk 2]
```diff
@@ -699,6 +704,11 @@ meshgen_block_cache_size (Mapblock mesh generator's MapBlock cache size MB) int
 #    Enables minimap.
 enable_minimap (Minimap) bool true
 
+#    Enables fullbright mode.
+#    This allows you to see in the dark if you have the required fullbright
+#    privilege.
+enable_fullbright_mode (Fullbright mode) bool false
```
Comment (paramat) [inline builtin/settingtypes.txt]: 'enable' is a redundant word that is unfortunately used for some other settings, please remove it.
fullbright mode = true / false is enough to make sense.

File: src/content_mapblock.cpp [hunk 3]
```diff
@@ -1000,7 +1012,10 @@ void MapblockMeshGenerator::drawPlantlikeRootedNode()
 	useTile(0, MATERIAL_FLAG_CRACK_OVERLAY, 0, true);
 	origin += v3f(0.0, BS, 0.0);
 	p.Y++;
-	if (data->m_smooth_lighting) {
+	bool fullbright = g_settings->getBool("fullbright_mode");
```
Comment (paramat) [inline src/content_mapblock.cpp]: This setting getter is too slow to use for every node drawn, fullbright mode needs to be cached in a bool.

File: src/content_mapblock.cpp [hunk 4]
```diff
@@ -1371,7 +1386,10 @@ void MapblockMeshGenerator::drawNode()
 			break;
 	}
 	origin = intToFloat(p, BS);
-	if (data->m_smooth_lighting)
+	bool fullbright = g_settings->getBool("fullbright_mode");
```
Comment (paramat) [inline src/content_mapblock.cpp]: As above.

File: src/mapblock_mesh.cpp [hunk 5]
```diff
@@ -850,7 +850,13 @@ static void getTileInfo(
 			layer.material_flags |= MATERIAL_FLAG_BACKFACE_CULLING;
 	}
 
-	if (!data->m_smooth_lighting) {
+	bool fullbright = g_settings->getBool("fullbright_mode");
```
Comment (paramat) [inline src/mapblock_mesh.cpp]: This setting getter is too slow to use for every tile, fullbright mode needs to be cached in a bool.

File: src/particles.cpp [hunk 6]
```diff
@@ -185,11 +185,16 @@ void Particle::updateLight()
 	else
 		light = blend_light(m_env->getDayNightRatio(), LIGHT_SUN, 0);
 
-	u8 m_light = decode_light(light + m_glow);
-	m_color.set(255,
-		m_light * m_base_color.getRed() / 255,
-		m_light * m_base_color.getGreen() / 255,
-		m_light * m_base_color.getBlue() / 255);
+	bool fullbright = g_settings->getBool("fullbright_mode");
```
Comment (paramat) [inline src/particles.cpp]: This setting getter is too slow to use here, fullbright mode needs to be cached in a bool.

**Comment by paramat** [comment]
> I don't know how to efficiently pass the fullbright setting (which is changed now and then) to the light calculation

See line comments, It's essential the setting is cached as a bool to not affect rendering performance. The setting getter is far too slow to use where you have used it.
:-1: for current code.

**Comment by paramat** [comment]
No progress or requested changes for 6 months, closing for now.

**Comment by Tcll** [comment]
this would be a really good thing to have
Minecraft has a potion of night vision which allows you to see in the dark for an amount of time
and of course there are numerous mods which allow enchanting a helmet or something of the sort to grant constant night vision

and then of course Aroma1997 has the mining world mod which is constantly lit if enabled in the settings

you guys saying things like fullbright has no use makes me think you guys have never actually played Minecraft :/
because there's tons of uses for this
heck even the option to set the world to have Minecraft lighting instead of pitch black caves would at least improve things
and no, this isn't cheating, Minecraft gets along just fine with it.

if you're afraid of cheating, add a mob engine

**Comment by sfan5** [comment]
This PR adds a fullbright mode for debugging purposes, not gameplay-controlled brightness like you are talking about.

**Comment by Tcll** [comment]
@sfan5 oh woops, I responded to the wrong issue, let me move that to the proper issue

**Comment by Tcll** [comment]
@sfan5 I take it you didn't see my post noting I accidentally posted here and would move my last post to it's appropriate issue. :/
I'll remove this post whenever you remove yours
thanks for informing me

**Pull request closed without merging**